src/portfft/dispatcher/global_dispatcher.hpp

```diff
@@ -306,57 +308,94 @@ struct committed_descriptor<Scalar, Domain>::run_kernel_struct<Dir, LayoutIn, La
-              twiddles_ptr, factors_and_scan, scale_factor, intermediate_twiddles_offset, impl_twiddle_offset, 0,
-              committed_size, static_cast<Idx>(max_batches_in_l2), static_cast<IdxGlobal>(num_batches),
-              static_cast<IdxGlobal>(i), static_cast<Idx>(factor_num), dimension_data.num_factors, l2_events,
-              desc.queue);
-          intermediate_twiddles_offset +=
-              2 * kernels.at(factor_num).batch_size * static_cast<IdxGlobal>(kernels.at(factor_num).length);
-          impl_twiddle_offset += detail::increment_twiddle_offset(kernels.at(factor_num).level,
-                                                                  static_cast<Idx>(kernels.at(factor_num).length));
+              current_kernel, desc.scratch_ptr_1.get(), desc.scratch_ptr_1.get(),
+              desc.scratch_ptr_1.get() + imag_offset, desc.scratch_ptr_1.get() + imag_offset, twiddles_ptr,
+              factors_and_scan, scale_factor, intermediate_twiddles_offset, impl_twiddle_offset, 0, committed_size,
+              static_cast<Idx>(max_batches_in_l2), static_cast<IdxGlobal>(num_batches), static_cast<IdxGlobal>(i),
+              static_cast<Idx>(factor_num), dimension0.num_factors, storage, l2_events, desc.queue);
+          intermediate_twiddles_offset += 2 * current_kernel.batch_size * static_cast<IdxGlobal>(current_kernel.length);
+          impl_twiddle_offset +=
+              detail::increment_twiddle_offset(current_kernel.level, static_cast<Idx>(current_kernel.length));
         }
+#ifdef PORTFFT_LOG_DUMPS
+        desc.queue.copy(desc.scratch_ptr_1.get(), tmp.data(), tmp.size(), l2_events).wait();
+        std::cout << std::endl;
+        std::cout << "after factor " << factor_num << ": ";
+        for (Scalar val : tmp) {
+          std::cout << val << ", ";
+        }
+        std::cout << std::endl;
+        std::cout << std::endl;
+#endif
       }
       event = desc.queue.submit([&](sycl::handler& cgh) {
         cgh.depends_on(l2_events);
         cgh.host_task([&]() {});
       });
+      event.wait();
       for (Idx num_transpose = num_transposes - 1; num_transpose > 0; num_transpose--) {
         event = detail::transpose_level<Scalar, Domain>(
             kernels.at(static_cast<std::size_t>(num_transpose) + static_cast<std::size_t>(num_factors)),
-            static_cast<const Scalar*>(desc.scratch_ptr_1.get()), desc.scratch_ptr_2.get(), factors_and_scan,
-            committed_size, static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_transpose,
-            num_factors, 0, desc.queue, desc.scratch_ptr_1, desc.scratch_ptr_2, {event});
+            desc.scratch_ptr_1.get(), desc.scratch_ptr_2.get(), factors_and_scan, committed_size,
+            static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_factors, 0, desc.queue,
+            {event}, storage);
         event.wait();
+        if (storage == complex_storage::SPLIT_COMPLEX) {
```

Comment:
You had figured out the issue with why that explicit `event.wait()` was required,( because of the kernel being enqueued before the pointer swap ). Are you still planning on fixing that ?
If not it's alright, and I can bundle that up with the Bluestein PR

Reply:
I already changed what I thought would fix that, but removing the waits still fails a limited set of tests.